File: surface/widgets/compose/components/widgets/form-controls/MultiSelect.kt
// ============================================================
// Clef Surface Compose Widget — MultiSelect
//
// Dropdown multi-choice selector. Displays a trigger showing
// the count of selected items with a dropdown containing
// checkbox-style items. Maps the multi-select.widget anatomy
// (root, label, trigger, content, item, itemIndicator) to
// Material 3 ExposedDropdownMenuBox with Checkbox menu items.
// ============================================================

package clef.surface.compose.components.widgets.formcontrols

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Checkbox
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.MenuAnchorType
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp

// --------------- Types ---------------

data class MultiSelectOption(
    val label: String,
    val value: String,
    val disabled: Boolean = false,
)

// --------------- Component ---------------

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MultiSelect(
    value: List<String>,
    options: List<MultiSelectOption>,
    onValueChange: (List<String>) -> Unit,
    modifier: Modifier = Modifier,
    label: String? = null,
    enabled: Boolean = true,
) {
    var expanded by remember { mutableStateOf(false) }

    val displayText = remember(value, options) {
        when {
            value.isEmpty() -> "Select..."
            value.size == 1 -> options.find { it.value == value[0] }?.label ?: value[0]
            else -> "${value.size} selected"
        }
    }

    Column(modifier = modifier) {
        if (label != null) {
            Text(
                text = label,
                style = MaterialTheme.typography.labelLarge,
            )
        }

        ExposedDropdownMenuBox(
            expanded = expanded,
            onExpandedChange = { if (enabled) expanded = it },
        ) {
            OutlinedTextField(
                value = displayText,
                onValueChange = {},
                readOnly = true,
                enabled = enabled,
                singleLine = true,
                trailingIcon = {
                    ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded)
                },
                modifier = Modifier
                    .menuAnchor(MenuAnchorType.PrimaryNotEditable)
                    .fillMaxWidth(),
            )

            ExposedDropdownMenu(
                expanded = expanded,
                onDismissRequest = { expanded = false },
            ) {
                options.forEach { option ->
                    val isChecked = option.value in value
                    val isEnabled = enabled && !option.disabled

                    DropdownMenuItem(
                        text = {
                            Row(verticalAlignment = Alignment.CenterVertically) {
                                Checkbox(
                                    checked = isChecked,
                                    onCheckedChange = null,
                                    enabled = isEnabled,
                                )
                                Spacer(modifier = Modifier.width(8.dp))
                                Text(
                                    text = option.label,
                                    color = if (isEnabled) {
                                        MaterialTheme.colorScheme.onSurface
                                    } else {
                                        MaterialTheme.colorScheme.onSurface.copy(alpha = 0.38f)
                                    },
                                )
                            }
                        },
                        onClick = {
                            if (isEnabled) {
                                val next = if (isChecked) {
                                    value - option.value
                                } else {
                                    value + option.value
                                }
                                onValueChange(next)
                            }
                        },
                        enabled = isEnabled,
                    )
                }
            }
        }
    }
}

File: surface/widgets/winui/components/widgets/form-controls/MultiSelect.cs
// ============================================================
// Clef Surface WinUI Widget — MultiSelect
//
// Dropdown multi-choice selector. Displays a trigger showing
// the count of selected items with a dropdown containing
// checkbox-style items. Maps the multi-select.widget anatomy
// to WinUI 3 ComboBox with CheckBox items.
// ============================================================

using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using System.Collections.Generic;
using System.Linq;

namespace Clef.Surface.WinUI.Widgets.FormControls;

public record MultiSelectOption(string Label, string Value, bool Disabled = false);

public sealed class ClefMultiSelect : UserControl
{
    public static readonly DependencyProperty LabelProperty =
        DependencyProperty.Register(nameof(Label), typeof(string), typeof(ClefMultiSelect),
            new PropertyMetadata(null, OnPropertyChanged));

    public string Label { get => (string)GetValue(LabelProperty); set => SetValue(LabelProperty, value); }

    public event System.EventHandler<List<string>> ValueChanged;

    private List<MultiSelectOption> _options = new();
    private List<string> _selectedValues = new();
    private readonly StackPanel _root;
    private readonly TextBlock _labelBlock;
    private readonly DropDownButton _trigger;
    private readonly Flyout _flyout;
    private readonly StackPanel _flyoutPanel;

    public ClefMultiSelect()
    {
        _labelBlock = new TextBlock { Visibility = Visibility.Collapsed, FontWeight = Microsoft.UI.Text.FontWeights.SemiBold };
        _flyoutPanel = new StackPanel { Spacing = 4, MinWidth = 200 };
        _flyout = new Flyout { Content = _flyoutPanel };
        _trigger = new DropDownButton { Content = "Select...", Flyout = _flyout };
        _root = new StackPanel { Spacing = 4 };
        _root.Children.Add(_labelBlock);
        _root.Children.Add(_trigger);
        Content = _root;
    }

    public void SetOptions(List<MultiSelectOption> options, List<string> selectedValues)
    {
        _options = options;
        _selectedValues = selectedValues.ToList();
        UpdateVisual();
    }

    private static void OnPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        => ((ClefMultiSelect)d).UpdateVisual();

    private void UpdateVisual()
    {
        _labelBlock.Text = Label ?? "";
        _labelBlock.Visibility = string.IsNullOrEmpty(Label) ? Visibility.Collapsed : Visibility.Visible;
        _trigger.Content = _selectedValues.Count switch
        {
            0 => "Select...",
            1 => _options.FirstOrDefault(o => o.Value == _selectedValues[0])?.Label ?? _selectedValues[0],
            _ => $"{_selectedValues.Count} selected"
        };
        _flyoutPanel.Children.Clear();
        foreach (var option in _options)
        {
            var cb = new CheckBox
            {
                Content = option.Label,
                IsChecked = _selectedValues.Contains(option.Value),
                IsEnabled = !option.Disabled
            };
            var val = option.Value;
            cb.Checked += (s, e) => { if (!_selectedValues.Contains(val)) _selectedValues.Add(val); OnSelectionChanged(); };
            cb.Unchecked += (s, e) => { _selectedValues.Remove(val); OnSelectionChanged(); };
            _flyoutPanel.Children.Add(cb);
        }
    }

    private void OnSelectionChanged()
    {
        _trigger.Content = _selectedValues.Count switch
        {
            0 => "Select...",
            1 => _options.FirstOrDefault(o => o.Value == _selectedValues[0])?.Label ?? _selectedValues[0],
            _ => $"{_selectedValues.Count} selected"
        };
        ValueChanged?.Invoke(this, _selectedValues.ToList());
    }
}
